File: leetcode-js/84.js
/**
 * @param {number[]} heights
 * @return {number}
 */
var largestRectangleArea = function (heights) {

    let stack = [];
    let maxArea = 0;

    for (let i = 0; i <= heights.length; i++) {
        let currHeight = (i === heights.length) ? 0 : heights[i];
        while (stack.length > 0 && currHeight < heights[stack[stack.length - 1]]) {
            let height = heights[stack.pop()];
            let width = stack.length === 0 ? i : i - stack[stack.length - 1] - 1;
            maxArea = Math.max(maxArea, height * width);
        }
        stack.push(i);
    }

    return maxArea;
};

let heights = [2, 1, 5, 6, 2, 3];
console.log(largestRectangleArea(heights));

File: leetcode-go/84.go
package main

func Max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func largestRectangleArea(heights []int) int {
	stack := []int{0}
	maxArea := 0

	for index := 0; index <= len(heights); index++ {
		currHeight := 0
		if index != len(heights) {
			currHeight = heights[index]
		}
		for len(stack) > 0 && currHeight < heights[stack[len(stack)-1]] {
			width := index
			height := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if len(stack) > 0 {
				width = index - stack[len(stack)-1] - 1
			}
			maxArea = Max(maxArea, height*width)
		}
		stack = append(stack, index)
	}

	return maxArea
}
